ext/drreg/drreg.c

```diff
@@ -1217,9 +1233,11 @@ drreg_statelessly_restore_app_value(void *drcontext, instrlist_t *ilist, reg_id_
          */
 #ifdef X86
     if (reg != DR_REG_NULL && pt->aflags.xchg == reg) {
-        pt->slot_use[AFLAGS_SLOT] = DR_REG_XAX; /* appease assert */
-        restore_reg(drcontext, pt, DR_REG_XAX, AFLAGS_SLOT, ilist, where_respill, false);
-        pt->slot_use[AFLAGS_SLOT] = DR_REG_NULL;
+        ASSERT(pt->aflags.slot != MAX_SPILLS, "Aflags slot not reserved");
+        pt->slot_use[pt->aflags.slot] = DR_REG_XAX; /* appease assert */
+        restore_reg(drcontext, pt, DR_REG_XAX, pt->aflags.slot, ilist, where_respill,
+                    false);
+        pt->slot_use[pt->aflags.slot] = DR_REG_NULL;
```

Comment:
Should this be calling reset_aflags_spill_slot?  If not please comment as to why it's different from the other cases.

Reply:
No, because we don't want to modify state in `drreg_statelessly_restore_app_value`; I also see that we pass `release=false` to `restore_reg` for the same reason probably.

Sure, I'll add a comment.

Reply:
But then slot_use and aflags.slot don't match.

But wait -- this is all inside "pt->aflags.xchg == reg" -- so is there a slot in the first place?  You added an assert there: does it not fire?  Looks like there should *not* be a slot, and it temporarily clobbers the slot to appease restore_reg asserts, and this set to null is just undoing that temp clobber.  I misread it as a true clear, which everywhere else uses the new helper.

Reply:
Something seems wrong: either the assert, or the code which clobbers what the assert is asserting.

Reply:
Yeah I was just looking at that to confirm -- it does seem weird. I checked that none of our current tests exercise this path.

Reply:
So it is possible for `pt->aflags.xchg == xax`  (aflags in xax) **and** `pt->aflags.slot != max_spills` (aflags spilled to slot) to be true at the same time. This happens as an intermediate state in `drreg_statelessly_restore_app_value`, when it is invoked for xax holding the spilled aflags value. When this happens, it invokes `drreg_move_aflags_from_reg` with `stateful=false`. This spills aflags to some slot at `where_restore`, but doesn't clear `pt->aflags.xchg`. And then in the `restore_reg` call, aflags is spilled again to xax at `where_respill`.

What we want for this PR is to simply add the extra book-keeping required for aflags slot -- change the `restore_reg` call to release the slot, and also call `reset_aflags_spill_slot`.

Added a test that invokes this path.